### PR TITLE
http2: using cached settings for multiplexing limits across the board

### PR DESCRIPTION
**Pull request by alyssawilk**
This turns up settings caching for HTTP/2 and ALPN upstreams, with each cluster caching the multiplexed streams for endpoints, and using those as the default for subsequent connections.

Risk Level: high (pool capacity math is super complicated)
Testing: prior integration testing, new unit tests
Docs Changes: n/a
Release Notes: inline
Runtime guard: envoy.reloadable_features.allow_concurrency_for_alpn_pool
Fixes https://github.com/envoyproxy/envoy/issues/20696

### CODE REVIEW

File: source/common/http/conn_pool_base.cc [hunk 1]
```diff
@@ -207,36 +207,13 @@ uint64_t MultiplexedActiveClientBase::maxStreamsPerConnection(uint64_t max_strea
   return (max_streams_config != 0) ? max_streams_config : DEFAULT_MAX_STREAMS;
 }
 
-MultiplexedActiveClientBase::MultiplexedActiveClientBase(HttpConnPoolImplBase& parent,
-                                                         uint32_t max_concurrent_streams,
-                                                         Stats::Counter& cx_total)
+MultiplexedActiveClientBase::MultiplexedActiveClientBase(
+    Envoy::Http::HttpConnPoolImplBase& parent, uint32_t effective_concurrent_streams,
```
Comment (RyanTheOptimist) [inline source/common/http/conn_pool_base.cc:211]: Out of curiosity, why the change from "HttpConnPoolImplBase" to "Envoy::Http::HttpConnPoolImplBase". This code is in the namespace Envoy::Http so I would have expected that to be redundant?

File: source/common/http/conn_pool_base.h [hunk 2]
```diff
@@ -88,22 +88,30 @@ class HttpConnPoolImplBase : public Envoy::ConnectionPool::ConnPoolImplBase,
...
   Random::RandomGenerator& randomGenerator() { return random_generator_; }
 
-  virtual absl::optional<HttpServerPropertiesCache::Origin> origin() { return {}; }
+  virtual absl::optional<HttpServerPropertiesCache::Origin>& origin() { return origin_; }
   virtual Http::HttpServerPropertiesCacheSharedPtr cache() { return nullptr; }
 
 protected:
   friend class ActiveClient;
 
+  absl::optional<HttpServerPropertiesCache::Origin> origin_;
```
Comment (RyanTheOptimist) [inline source/common/http/conn_pool_base.h:97]: nit: This should be private: according to the style guide (which would suggest we need a protected setter.

File: source/common/http/conn_pool_base.h [hunk 3]
```diff
@@ -199,10 +198,9 @@ class MultiplexedActiveClientBase : public CodecClientCallbacks,
                                     public Http::ConnectionCallbacks,
                                     public Envoy::Http::ActiveClient {
 public:
-  MultiplexedActiveClientBase(HttpConnPoolImplBase& parent, uint32_t max_concurrent_streams,
-                              Stats::Counter& cx_total);
-  MultiplexedActiveClientBase(HttpConnPoolImplBase& parent, uint32_t max_concurrent_streams,
-                              Stats::Counter& cx_total, Upstream::Host::CreateConnectionData& data);
+  MultiplexedActiveClientBase(HttpConnPoolImplBase& parent, uint32_t effective_concurrent_streams,
+                              uint32_t max_configured_concurrent_streams, Stats::Counter& cx_total,
+                              OptRef<Upstream::Host::CreateConnectionData> data);
```
Comment (RyanTheOptimist) [inline source/common/http/conn_pool_base.h:203]: Thanks for cleaning this up! It's annoying when there are N diff arguments to a constructor an N! various constructors.

File: source/common/http/http1/conn_pool.cc [hunk 4]
```diff
@@ -74,16 +74,16 @@ void ActiveClient::StreamWrapper::onResetStream(StreamResetReason, absl::string_
 ActiveClient::ActiveClient(HttpConnPoolImplBase& parent)
     : Envoy::Http::ActiveClient(
           parent, parent.host()->cluster().maxRequestsPerConnection(),
-          1 // HTTP1 always has a concurrent-request-limit of 1 per connection.
-      ) {
+          1, // HTTP1 always has a concurrent-request-limit of 1 per connection.
+          1, absl::nullopt) {
```
Comment (RyanTheOptimist) [inline source/common/http/http1/conn_pool.cc:78]: nit would you mind adding comments (or /*effective_concurrent_stream_limit=*/, etc) to the literals?

File: test/common/http/mixed_conn_pool_test.cc [hunk 5]
```diff
@@ -106,11 +106,19 @@ void MixedConnPoolImplTest::testAlpnHandshake(absl::optional<Protocol> protocol)
   conn_pool_.reset();
 }
 
-TEST_F(MixedConnPoolImplTest, BasicNoAlpnHandshake) {
+TEST_F(MixedConnPoolImplTest, BasicNoAlpnHandshakeOld) {
```
Comment (RyanTheOptimist) [inline test/common/http/mixed_conn_pool_test.cc:109]: Do we need an h3 pool test too?

**Review by RyanTheOptimist — COMMENTED** [review]
LOVE THIS!

File: source/common/http/conn_pool_base.cc (hunk 1 above)
Comment (alyssawilk) [reply source/common/http/conn_pool_base.cc:211]: oh, that was from the "deleted" constructor below - git diff weirdness.  fixed either way

File: source/common/http/conn_pool_base.h (hunk 3 above)
Comment (alyssawilk) [reply source/common/http/conn_pool_base.h:203]: oh my gosh it was so hard adding my argument so I just gave up and changed all the things :-P

File: test/common/http/mixed_conn_pool_test.cc (hunk 5 above)
Comment (alyssawilk) [reply test/common/http/mixed_conn_pool_test.cc:109]: I don't think so - there's no changes to the H3 pool

**Review by mattklein123 — APPROVED** [review]
Nice!!
